extension/basicauthextension/extension.go

```diff
@@ -38,50 +39,54 @@ var (
 )
 
 type basicAuth struct {
-	htpasswd  HtpasswdSettings
-	matchFunc func(username, password string) bool
+	htpasswd     HtpasswdSettings
+	matchFunc    func(username, password string) bool
+	userPassPair string
 }
 
-func newExtension(cfg *Config) (configauth.ServerAuthenticator, error) {
+func newExtension(cfg *Config) (*basicAuth, error) {
```

Comment:
I think I would prefer the factory to determine whether it needs a client authenticator or a server authenticator (see `createExtension` in the factory.go).

I also think this extension has two different sets of configuration options:

- username and password for the client authenticator, with only one subject
- htpasswd for the server authenticator, with potentially multiple subjects

So, if both the username and password are set, this takes the shape of a client authenticator, otherwise it's a server authenticator. The extension cannot be both at the same time. As an operator, I would find it confusing to have one extension instance with the two facets at the same time.

Reply:
Good points. I have incorporated that logic. I have also enforced that this extention will only act as either a client auth or server auth. If a user needs both, there should be two. Kindly review.